code/__DATASTRUCTURES/priority_queue.dm

```diff
@@ -0,0 +1,65 @@
+
+//////////////////////
+//PriorityQueue object
+//////////////////////
+
+//an ordered list, using the cmp proc to weight the list elements
+/PriorityQueue
+	var/list/L //the actual queue
+	var/cmp //the weight function used to order the queue
+
+/PriorityQueue/New(compare)
+	L = new()
+	cmp = compare
+
+/PriorityQueue/proc/IsEmpty()
+	return !L.len
+
+//add an element in the list,
+//immediatly ordering it to its position using Insertion sort
+/PriorityQueue/proc/Enqueue(atom/A)
+	var/i
+	L.Add(A)
+	i = L.len -1
+	while(i > 0 &&  call(cmp)(L[i],A) >= 0) //place the element at it's right position using the compare proc
+		L.Swap(i,i+1) 						//last inserted element being first in case of ties (optimization)
+		i--
+
+//removes and returns the first element in the queue
+/PriorityQueue/proc/Dequeue()
+	if(!L.len)
```

Comment:
Making an extra function call is a performance trade-off with little to no gain in this case.